fe/fe-core/src/main/java/org/apache/doris/nereids/rules/rewrite/PredicatePropagation.java

```diff
@@ -70,19 +72,44 @@ public Expression visit(Expression expr, Void context) {
 
             @Override
             public Expression visitComparisonPredicate(ComparisonPredicate cp, Void context) {
-                if (cp.left().isSlot() && cp.right().isConstant()) {
-                    return replaceSlot(cp);
-                } else if (cp.left().isConstant() && cp.right().isSlot()) {
-                    return replaceSlot(cp);
+                if (isExpressionSlotCoveredByCast(cp.left()) && (cp.right().isConstant())) {
+                    return replaceSlot(cp, getDatatypeCoveredByCast(cp.left()));
+                } else if (isExpressionSlotCoveredByCast(cp.right()) && cp.left().isConstant()) {
+                    return replaceSlot(cp, getDatatypeCoveredByCast(cp.right()));
                 }
                 return super.visit(cp, context);
             }
 
-            private Expression replaceSlot(Expression expr) {
+            private boolean isTwoExpressionEqualWithCast(Expression left, Expression right) {
+                return getSlotCoveredByCast(left).equals(getSlotCoveredByCast(right));
+            }
+
+            private boolean canCompareType(DataType dataType) {
+                return dataType.isBigIntType() || dataType.isIntegerType() || dataType.isSmallIntType()
+                        || dataType.isTinyIntType() || dataType.isLargeIntType();
+            }
+
+            private boolean isOriginDataTypeBigger(DataType originDataType, Expression expr) {
+                if (canCompareType(leftSlotEqualToRightSlot.child(0).getDataType())
+                        && canCompareType(leftSlotEqualToRightSlot.child(1).getDataType())
+                                && canCompareType(originDataType)) {
+                    // infer filter can not be lower than original datatype, or dataset would be wrong
+                    if (originDataType.width() > leftSlotEqualToRightSlot.child(0).getDataType().width()
+                            || originDataType.width() > leftSlotEqualToRightSlot.child(1).getDataType().width()) {
```

Comment:
use width not a good idea. maybe u should add widerThan or similar method to IntegralType

Reply:
done